pkg/policy/bucket-policy.go

```diff
@@ -82,6 +84,33 @@ type User struct {
 	CanonicalUser set.StringSet `json:"CanonicalUser,omitempty"`
 }
 
+// UnmarshalJSON is a custom json unmarshaler for Principal field,
+// the reason is that Principal can take a json struct represented by
+// User string but it can also take a string.
+func (u *User) UnmarshalJSON(data []byte) error {
+	// Try to unmarshal data in a struct equal to User, we need it
+	// to avoid infinite recursive call of this function
+	type AliasUser User
+	var au AliasUser
+	err := json.Unmarshal(data, &au)
+	if err == nil {
+		*u = User(au)
+		return nil
+	}
+	// Data type is not known, check if it is a json string
+	// which contains a star, which is permitted in the spec
+	var str string
+	err = json.Unmarshal(data, &str)
+	if err == nil {
+		if str != "*" {
+			return errors.New("unrecognized Principal field")
```

Comment:
Is it only `*` allowed by AWS?  If not, we have to support it to preserve policy rules for updates.

Reply:
> Is it only * allowed by AWS?

I guess so, nothing indicates it in this link https://docs.aws.amazon.com/AmazonS3/latest/dev/s3-bucket-user-policy-specifying-principal-intro.html

Reply:
then we have to support all values not limiting to `*`

Reply:
I tested with `aws s3api put-bucket-policy`, AWS responds with "Invalid bucket policy syntax" for whatever value we put, it accepts only `*`.  So I guess it is fine.